crates/types/src/spline_motion_interpolator.rs

```diff
@@ -0,0 +1,107 @@
+use splines::{Interpolation, Key, Spline};
+use thiserror::Error;
+
+use crate::{Joints, MotionFile};
+use std::time::Duration;
+
+pub struct SplineMotionFileInterpolator {
+    interpolator: Spline<f32, Joints>,
+    current_time: Duration,
+    start_time: Duration,
+    end_time: Duration,
+}
+
+#[derive(Error, Debug)]
+pub enum InterpolatorError {
+    #[error("accessed interpolator defined for [{min}s, {max}s] at {current}s")]
+    OutOfBoundsAccess { min: f32, max: f32, current: f32 },
+}
+
+impl From<MotionFile> for SplineMotionFileInterpolator {
+    fn from(motion_file: MotionFile) -> Self {
+        assert!(!motion_file.frames.is_empty());
+
+        let mut current_time = Duration::ZERO;
+        let mut keys = vec![Key::new(
+            current_time,
+            motion_file.initial_positions,
+            Interpolation::Linear,
+        )];
+
+        keys.extend(motion_file.frames.into_iter().map(|frame| {
+            current_time += frame.duration;
+            Key::new(current_time, frame.positions, Interpolation::Linear)
+        }));
+
+        Self::new(keys)
+    }
+}
+
+impl SplineMotionFileInterpolator {
+    pub fn new(keys: Vec<Key<Duration, Joints>>) -> Self {
+        assert!(keys.len() >= 2, "need at least two keys to interpolate");
+        //NOTE: assume keys are sorted
+        let last_key_index = keys.len() - 1;
+
+        let start_time = keys[0].t;
+        let current_time = start_time;
+        let end_time = keys[last_key_index].t;
+
+        let left_helper = Key::new(
+            2. * keys[0].t.as_secs_f32() - keys[1].t.as_secs_f32(),
+            keys[1].value,
+            Interpolation::Linear,
+        );
+        let right_helper = Key::new(
+            2. * keys[last_key_index].t.as_secs_f32() - keys[last_key_index - 1].t.as_secs_f32(),
+            keys[last_key_index - 1].value,
+            Interpolation::Linear,
+        );
```

Comment:
What actually happens here? What are the helpers used for? Could you extract some code into methods to document it via method names?

Reply:
Extracted into method